### PR TITLE
Viet/balance tx

### PR DESCRIPTION
**Pull request by vvtran**
- Balancing Tx https://github.com/Plutonomicon/cardano-browser-tx/issues/7
- Various Eq, Ord and Newtype instances.
- Small javascript changes.
- This has been rebased with Rory's nixification.
- Includes `Eq/Ord` changes for ByteArrays -> TokenName and CurrencySymbol from https://github.com/Plutonomicon/cardano-browser-tx/pull/51
- Includes changes from https://github.com/Plutonomicon/cardano-browser-tx/pull/31
- Rewrite of prebalancer from https://github.com/mlabs-haskell/bot-plutus-interface/blob/master/src/BotPlutusInterface/PreBalance.hs along with a post Ada-only balancer.
- Copied helpers for Value and Ada from Plutus.
- Rewrote Value slightly so that the (non-Ada) Map is a newtype so we can write custom monoid instance.

### CODE REVIEW

**Comment by vvtran** [comment]
Converting to draft.

**Comment by Benjmhart** [comment]
are you running an autoformatter? there's a bunch of diffs here not relevant to the task, perhaps run autoformat in a separate PR to make review easier.

**Comment by vvtran** [comment]
> are you running an autoformatter? there's a bunch of diffs here not relevant to the task, perhaps run autoformat in a separate PR to make review easier.

I have a whitespace remover, and the exports sometime rearrange themselves if I don't write them out manually. Otherwise, I tend to just reformat stuff as I read code. I'll try to refrain in the future if it makes  reviewing difficult.

Also, this PR isn't quite ready yet, I still need to translate some of the effects & cardano CLI functions across to our monad stack.

**Comment by Benjmhart** [comment]
No sweat

**Comment by vvtran** [comment]
@Benjmhart  Could I get some feedback on this please? It's by no means ready to be merged and relies on other tickets so there's a fair bit of placeholder code. I used a whitespace remover to edit various files to give redundant changes. So to make review easier, the following files should get checked (at least):
- `BalanceTx`
- `ProtocolParametersAlonzo`
- `Types.Ada` (this is pretty much copy and paste from Plutus)
- `Types.Value` (this is pretty much copy and paste from Plutus)
- `Types.Transaction`
I merged a few changes from https://github.com/Plutonomicon/cardano-browser-tx/pull/31 as I went along so you can ignore these.

It follows the prebalancer logic from https://github.com/mlabs-haskell/mlabs-pab/blob/master/src/MLabsPAB/PreBalance.hs to get into a state where the input value is greater or equal than the output (which includes min ada utxos) + fees; this is essentially my `preBalanceTxBody` function. Therefore, all excess will be in Ada that requires returning to the owner. In Mlabs pab, the excess Ada is balanced by calling a Cardano CLI build transaction function which convenient builds the final transaction too - here I've created my own function called returnAdaChange.

File: src/Types/Ada.purs
Comment (ngua) [inline src/Types/Ada.purs:55]: I suppose this line should be removed :)

File: src/BalanceTx.purs
Comment (ngua) [inline src/BalanceTx.purs:63]: I think it might be a good idea to make a more specific error type (applies to the whole module)

**Comment by ngua** [comment]
@vvtran Can you run `make autogen-deps` and commit the resulting autogenerated files please (there should be a new `spago-packages.nix`)?

**Comment by ngua** [comment]
Also, I think it might be a good idea to rebase onto/merge https://github.com/klntsky/cardano-browser-tx/tree/klntsky/fix-tokenname-representation and then make the necessary adjustments (especially `adaToken` and `adaValue`)

File: src/Types/JsonWsp.purs
Comment (Benjmhart) [inline src/Types/JsonWsp.purs:249]: so we need a ticket to support native assets right?   let's prefer tickets over TODO's

File: src/Types/Transaction.purs
Comment (Benjmhart) [inline src/Types/Transaction.purs:122]: just thinking for all these instances...  shouldn't they be newtype instances?

Comment (Benjmhart) [reply src/Types/Transaction.purs:122]: i believe this reduces code size.

File: src/BalanceTx.purs
Comment (ngua) [inline src/BalanceTx.purs:602]: We should probably open another issue about this

**Comment by vvtran** [comment]
> @vvtran Can you run `make autogen-deps` and commit the resulting autogenerated files please (there should be a new `spago-packages.nix`)?

@ngua is it weird that there is nothing to commit?

```
viet@viet-ThinkPad-E15-Gen-2:~/Work/MLabs/cardano-browser-tx$ make autogen-depsspago2nix generate \
	&& node2nix -l package-lock.json \
	&& mv default.nix node2nix.nix \
	&& git restore default.nix
getting packages..
got 89 packages from spago ls deps.
fetching aff
fetching ansi
fetching argonaut
fetching argonaut-codecs
fetching argonaut-core
fetching argonaut-traversals
fetching arraybuffer-types
fetching arrays
fetching avar
fetching bifunctors
fetching bigints
fetching catenable-lists
fetching console
fetching const
fetching contravariant
fetching control
fetching datetime
fetching debug
fetching distributive
fetching effect
fetching either
fetching enums
fetching exceptions
fetching exists
fetching foldable-traversable
fetching foreign
fetching foreign-object
fetching fork
fetching free
fetching freet
fetching functions
fetching functors
fetching gen
fetching identity
fetching integers
fetching invariant
fetching js-date
fetching lazy
fetching lcg
fetching lists
fetching math
fetching maybe
fetching medea
fetching mmorph
fetching mote
fetching naturals
fetching newtype
fetching node-buffer
fetching node-fs
fetching node-fs-aff
fetching node-path
fetching node-streams
fetching nonempty
fetching now
fetching nullable
fetching numbers
fetching ordered-collections
fetching orders
fetching parallel
fetching parsing
fetching partial
fetching pipes
fetching prelude
fetching profunctor
fetching profunctor-lenses
fetching psci-support
fetching quickcheck
fetching quickcheck-combinators
fetching quickcheck-laws
fetching random
fetching record
fetching refs
fetching safe-coerce
fetching safely
fetching spec
fetching st
fetching strings
fetching tailrec
fetching these
fetching transformers
fetching tuples
fetching type-equality
fetching typelevel
fetching typelevel-prelude
fetching undefined
fetching unfoldable
fetching unicode
fetching unordered-collections
fetching unsafe-coerce
wrote spago-packages.nix
fetching local directory: ./. from .

WARNING: There is a node_modules/ folder in the root directory of the project!
These packages will be included in the Nix build and influence the outcome.
If you don't want this to happen, then you should remove it before running any
of the Nix commands!
```
and running `git status` says my working tree is clean.

File: src/Types/Transaction.purs
Comment (vvtran) [reply src/Types/Transaction.purs:122]: Makes sense, I will look to change them all.

File: src/Types/Ada.purs
Comment (vvtran) [reply src/Types/Ada.purs:55]: Ah, the classic.

File: src/BalanceTx.purs
Comment (vvtran) [reply src/BalanceTx.purs:63]: Great shout.

**Comment by ngua** [comment]
Ah, I see, my bad. I saw that `spago.dhall` had been changed but that there was no new `spago-packages.nix`. I thought that new packages had been added, but I guess it was just a dhall formatter or something sorting the list

**Comment by vvtran** [comment]
@Benjmhart  @ngua 
- Different deriving instances, more newtypes where possible (more on this later).
- As per suggestion, I've merged in https://github.com/klntsky/cardano-browser-tx/tree/klntsky/fix-tokenname-representation  (and later a non-forked branch) which is the uint8Array representation of token names and currencysymbol. As a result, I had to rewrite a few helper functions and instances (outside of Show), e.g. `Eq` and `Ord` which I initially wrote but Vladmir was also working in parallel on this so I ended up remerging his changes in. I initially created a `UInt8Array` module which has been deprecated in favour of Vladmir's `Types.Bytearrays`.
- I made imports for data types and data constructors (at least for the modules I'm working on) explicit.
- @Benjmhart , one of your comments about adding a ticket for TODO wasn't originally my code (I think it got caught up in the white space remover), but I've created a ticket anyhow, please check it isn't missing anything - see ticket #46.
- I've also created various other tickets on the backend of writing the balancer, tickets 47-50.
- As per another suggestion, I've changed the balancer error from strings to data types (per helper function) so they're modular and heterogeneous. The balancer then uses lmap to make these errors homogeneous - hopefully, this provides better type safety.
- Added more type signatures into helper functions in Value to assist readability.

It's likely that Vladmir will make further changes to his `Types.Bytearrays` module so I will merge those as required.

File: package.json
Comment (ngua) [inline package.json:21]: Could you remove these please? Perhaps this is from an outdated version of @adamczykm branch. These should all be provided by Nix

**Review by ngua — CHANGES_REQUESTED** [review]
The balancing logic looks good to me and I think everything has been resolved from previous comments. Beyond a few merge conflicts that need to be resolved, I think you should merge/rebase onto @adamczykm's branch again

File: package.json
Comment (vvtran) [reply package.json:22]: Did you mean this branch? https://github.com/Plutonomicon/cardano-browser-tx/tree/adamczykm/addr1_conv it's apparently up to date.

**Comment by vvtran** [comment]
@ngua I've merged `master` into this branch, trying to compare conflicts from  https://github.com/Plutonomicon/cardano-browser-tx/tree/adamczykm/addr1_conv and `master` "reasonably". The safest thing to do is to let that branch merge in of course - I may need to look at conflicts again.

Also note that I made a change to `Types.Transactions`, replacing `UInt8Array` with `Types.ByteArray (ByteArray)` from https://github.com/Plutonomicon/cardano-browser-tx/pull/51 We can use our single source of truth for bytearray `Eq`. THis was required for balancing logic, along with various other `newtype eq` derivations I added to `Types.Transactions`.

File: src/Types/Value.purs
Comment (klntsky) [inline src/Types/Value.purs:47]: Both links 404'd

Comment (vvtran) [reply src/Types/Value.purs:47]: Ah, it's changed name, wonderful :D https://github.com/mlabs-haskell/bot-plutus-interface/blob/master/src/BotPlutusInterface/PreBalance.hs good spot.

**Comment by vvtran** [comment]
Just updated the link to https://github.com/mlabs-haskell/bot-plutus-interface/blob/master/src/BotPlutusInterface/PreBalance.hs will check on its conflict status tomorrow depending on if the other PRs merging.

**Review by ngua — CHANGES_REQUESTED** [review]
OK, everything looks good. You should just merge with @klntsky's `ByteArray` branch again (hopefully this is the last merge :))

**Review by ngua — APPROVED** [review]

**Comment by ngua** [comment]
Let's merge this after merge #51, just waiting for @adamczykm to approve the changes to that one

There are a few placeholders here and there for @adamczykm's addr1 conversion, but I think that's OK for now

File: src/BalanceTx.purs
Comment (klntsky) [inline src/BalanceTx.purs:188]: Instance names are optional in PureScript already

Comment (vvtran) [reply src/BalanceTx.purs:188]: @ngua something worth deciding for https://github.com/Plutonomicon/cardano-browser-tx/issues/56
I'm not sure what is considered idiomatic but I've always added names, although I should add I've only been learning for a few months.

Comment (ngua) [reply src/BalanceTx.purs:188]: As a personal preference, I would like to omit them. They add a lot of noise in my opinion. I think it's OK for now if there's some discrepancies between different modules when it comes to this, though. We can go back and make everything more uniform later

**Review by MatthewCroughan — APPROVED** [review]
As far as the Nix changes go, this is fine. Though I would worry about getting carried away with node2nix, and recommend using https://github.com/davhau/dream2nix instead, unless you're going to document how to use node2nix for other devs.

**Comment by vvtran** [comment]
I reckon we can get this merged if you're happy with this.

**Comment by vvtran** [comment]
@ngua there's been a few merges into `master` since I posted so if you could check up to the conflicts that I will deal with tomorrow, that'd be great.

**Comment by ngua** [comment]
@vvtran I'll take a look today

**Comment by vvtran** [comment]
Decided to merge some stuff from https://github.com/Plutonomicon/cardano-browser-tx/pull/73 over since I made some small changes that are worth adding in, will need to merge with master now

**Comment by vvtran** [comment]
Just merged in from `master` so should be conflict free now.

File: src/Types/Ada.purs
Comment (ngua) [inline src/Types/Ada.purs:1]: If this is only for `ProtocolParametersAlonzo`, I would prefer to get rid of this module. We already have the `Coin` type and it will get confusing having both, I think. There are some useful functions at the bottom we can just move to `Types.Value` to use with `Coin`

File: src/ProtocolParametersAlonzo.purs
Comment (ngua) [inline src/ProtocolParametersAlonzo.purs:24]: I think these could just be `BigInt`s

File: src/Types/Ada.purs
Comment (ngua) [inline src/Types/Ada.purs:47]: These are helpful, but I think these could just be `Coin` and moved into `Types.Value`

File: src/ProtocolParametersAlonzo.purs
Comment (ngua) [reply src/ProtocolParametersAlonzo.purs:24]: Or `Coin` preferably

File: src/BalanceTx.purs
Comment (ngua) [inline src/BalanceTx.purs:284]: Vladimir's implementation was merged, right?

**Review by ngua — COMMENTED** [review]
I don't want to block merging this since it's a big PR and we can just make issues later if we feel it's necessary. Things look good in general, but I would like more guarantees about how `Value`s are constructed and used. But we eventually need to ensure that users cannot insert "ada" into the `NonAdaAsset`. I would recommend:

- removing the `Newtype` instances for `TokenName`, `CurrencySymbol`, etc... and hiding their constructors
- only exporting smart constructors that return `Maybe`s

I do think it's good to check at any rate for the presence of spurious "ada" in the `NonAdaAsset` as you've done here. But that's pretty easily circumvented when all of the constructors are exported

I've left some other comments as well. The only additional change I'd personally really like to see is dropping `Ada` and just using `Coin` instead

File: src/BalanceTx.purs
Comment (vvtran) [reply src/BalanceTx.purs:284]: I was going to leave this for a second PR, just to reduce the amount of potentially breaking dependencies if it changes again.

File: src/Types/Value.purs
Comment (vvtran) [reply src/Types/Value.purs:249]: I will rethink this after resolving issues with exported constructors as the balancer relies on this ATM.

File: src/Types/Value.purs
Comment (vvtran) [reply src/Types/Value.purs:7]: Will created issues.

**Pull request merged**
